pkg/syncer/resourcesync/controller.go

```diff
@@ -365,79 +380,128 @@ func (c *Controller) checkSSAR(ctx context.Context, gvr schema.GroupVersionResou
 	return sar.Status.Allowed, nil
 }
 
-// stopUnusedSyncerInformers stop syncers for gvrs not in requiredGVRs.
-func (c *Controller) stopUnusedSyncerInformers(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) {
+// removeUnusedGVRs removes the GVRs which are not required anymore, and notifies subscribes
+// that the list of supported GVRs changed.
+func (c *controller) removeUnusedGVRs(ctx context.Context, requiredGVRs map[schema.GroupVersionResource]bool) error {
 	logger := klog.FromContext(ctx)
 
-	c.mutex.Lock()
-	defer c.mutex.Unlock()
+	c.gvrsToWatchLock.Lock()
+	defer c.gvrsToWatchLock.Unlock()
 
-	for gvr, informer := range c.syncerInformerMap {
+	updated := false
+	for gvr := range c.gvrsToWatch {
 		if _, ok := requiredGVRs[gvr]; !ok {
 			logger.WithValues("gvr", gvr.String()).V(2).Info("Stop syncer for gvr")
-			informer.cancel()
-			delete(c.syncerInformerMap, gvr)
+			delete(c.gvrsToWatch, gvr)
+			updated = true
+		}
+	}
+
+	if updated {
+		c.notifySubscribers(ctx)
```

Comment:
don't do this within the lock

Reply:
Fixed in commit da89d306186c852a5f9f9ab2a3a549f896f5f7e0